### ISSUE TITLE
Can't override rust toolchain since #319153

### ISSUE THREAD

**Issue by bouk** [NixOS/nixpkgs#326112]
@Kranzes I'm running an issue with overriding the rust toolchain since #319153, I pass the following to crate2nix:

```nix
buildRustCrateForPkgs = pkgs: pkgs.buildRustCrate.override {
  cargo = toolchain;
  rustc = toolchain;
};
```

to set the toolchain when building something, since I don't want to override the toolchain globally. But after this PR it doesn't work because it uses the toolchain in `pkgsBuildBuild` instead of the one I'm passing in. Now I don't really know how to override the rust toolchain to the version I want.

`toolchain` is a nightly version I got from https://github.com/oxalica/rust-overlay

This problem is caused by this optimization:

https://github.com/NixOS/nixpkgs/blob/eec72fd9d15243e45e96fc5bc79b8edf927410d0/pkgs/build-support/rust/build-rust-crate/default.nix#L19-L27

Another issue is that the `rustc.llvmPackages` attribute is missing in rust-overlay:

https://github.com/NixOS/nixpkgs/blob/eec72fd9d15243e45e96fc5bc79b8edf927410d0/pkgs/build-support/rust/build-rust-crate/build-crate.nix#L30-L31

**Comment by jrobsonchase** [comment]
I'm having this issue as well. I've overridden my `nixpkgs` input to my flake to b2852eb9365c6de48ffb0dc2c9562591f652242a as a workaround for now.

**Comment by Kranzes** [comment]
Hi, sorry for the delay, I'm back home now from "holiday". Can any of you give me a minimal flake that reproduces the issue so I can experiment and find the best way to fix it?

**Comment by Fuuzetsu** [comment]
I'm also hitting this. I don't have a minimal flake but I can at least offer the error message which seems to be missing from the OP.

```
       error: attribute 'unwrapped' missing
       at /nix/store/0qzq33hvd7194rbl6xkv5hs2yj4nhbj4-source/pkgs/build-support/rust/build-rust-crate/default.nix:22:63:
           21|     (lib.splitString "," (lib.removePrefix "--target=" (
           22|       lib.elemAt (lib.filter (f: lib.hasPrefix "--target=" f) pkgsBuildBuild.rustc.unwrapped.configureFlags) 0)
             |                                                               ^
           23|     ));

```

**Comment by Kranzes** [comment]
Let me know if this https://github.com/NixOS/nixpkgs/pull/339298 fixes it.

**Comment by bouk** [comment]
@Kranzes unfortunately it does not, because it still defaults to `pkgsBuildBuild.rustc` for the rustc compiler. 

I've created a reproduction here: https://github.com/bouk/nixpkgs-issue-326112

As you can see in flake.nix I'm overriding rustc and cargo to the specific nightly version I want, but when I run `nix run .#test` I get the following output:

```
Rust version: rustc 1.80.1 (3f5fd8dd4 2024-08-06) (built from a source tarball)
```

This is the rustc version that's current in nixpkgs, not the `2024-01-01` nightly I specified.

Thanks a lot for looking into this 🙏

**Comment by Kranzes** [comment]
@bouk One problem with your reproducer is that it uses the `lib.makeOverridable` for `buildRustCrate` which takes an arg called `rust` not `rustc`. I think that still wouldn't have solved it. I made some changes in the PR that would allow passing a custom toolchain via the `lib.makeOverridable` but also using `buildRustCrate.override`. I did some refactoring to get rid of the `rustc'` and `cargo'` variables.
